# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [21]:
# Import your libraries:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
import sklearn as sk
import seaborn as sns
import matplotlib.pyplot as plt
import scipy as sc
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [53]:
# Your code here:
import sklearn as sk
from sklearn.datasets import load_diabetes



In [54]:
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [55]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [56]:
# Your code here:

print(diabetes.DESCR)


.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
http://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Brad

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

### Enter your answer here:

1. There are 10 attributes or features. There are all features trying to explain diabetes progression. 
2. Since the features have been converted into quantitative scale and have been standardized, they are now in same scale as the target variables. This means the relationship is linear.
3. 442


#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [57]:
diabetes.data[0]

array([ 0.03807591,  0.05068012,  0.06169621,  0.02187235, -0.0442235 ,
       -0.03482076, -0.04340085, -0.00259226,  0.01990842, -0.01764613])

In [58]:
diabetes.target[0]

151.0

In [59]:
# Your code here:
diabetes.feature_names

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [60]:
diabetes_df = pd.DataFrame(diabetes.data)

In [61]:
diabetes_df.head()

0         1         2         3         4         5         6  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

          7         8         9  
0 -0.002592  0.019908 -0.017646  
1 -0.039493 -0.068330 -0.092204  
2 -0.002592  0.002864 -0.025930  
3  0.034309  0.022692 -0.009362  
4 -0.002592 -0.031991 -0.046641

In [62]:
diabetes_df.columns = diabetes.feature_names

In [63]:
diabetes_df["disease_prog"] = diabetes.target

In [64]:
diabetes_df.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  disease_prog  
0 -0.002592  0.019908 -0.017646         151.0  
1 -0.039493 -0.068330 -0.092204          75.0  
2 -0.002592  0.002864 -0.025930         141.0  
3  0.034309  0.022692 -0.009362         206.0  
4 -0.002592 -0.031991 -0.046641         135.0

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [65]:
X = diabetes_df.iloc[:,:-1]

In [66]:
X

age       sex       bmi        bp        s1        s2        s3  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
5   -0.092695 -0.044642 -0.040696 -0.019442 -0.068991 -0.079288  0.041277   
6   -0.045472  0.050680 -0.047163 -0.015999 -0.040096 -0.024800  0.000779   
7    0.063504  0.050680 -0.001895  0.066630  0.090620  0.108914  0.022869   
8    0.041708  0.050680  0.061696 -0.040099 -0.013953  0.006202 -0.028674   
9   -0.070900 -0.044642  0.039062 -0.033214 -0.012577 -0.034508 -0.024993   
10  -0.096328 -0.044642 -0.083808  0.008101 -0.103389 -0.090561 -0.013948   
11   0.027178  0.050680  0.017506 -0.033214 -0.007073  0.045972 -0.065491   
12   0.016281 -0.044642 -0.028840 -0.009113 -0.004321 -0.009769  0.044958   
13   0.005383  0.050680 -0.001895  0.008101 -0.004321 -0.015719 -0.002903   
14   0.045341 -0.044642 -0.025607 -0.012556  0.017694 -0.000061  0.081775   
15  -0.052738  0.050680 -0.018062  0.080401  0.089244  0.107662 -0.039719   
16  -0.005515 -0.044642  0.042296  0.049415  0.024574 -0.023861  0.074412   
17   0.070769  0.050680  0.012117  0.056301  0.034206  0.049416 -0.039719   
18  -0.038207 -0.044642 -0.010517 -0.036656 -0.037344 -0.019476 -0.028674   
19  -0.027310 -0.044642 -0.018062 -0.040099 -0.002945 -0.011335  0.037595   
20  -0.049105 -0.044642 -0.056863 -0.043542 -0.045599 -0.043276  0.000779   
21  -0.085430  0.050680 -0.022373  0.001215 -0.037344 -0.026366  0.015505   
22  -0.085430 -0.044642 -0.004050 -0.009113 -0.002945  0.007767  0.022869   
23   0.045341  0.050680  0.060618  0.031053  0.028702 -0.047347 -0.054446   
24  -0.063635 -0.044642  0.035829 -0.022885 -0.030464 -0.018850 -0.006584   
25  -0.067268  0.050680 -0.012673 -0.040099 -0.015328  0.004636 -0.058127   
26  -0.107226 -0.044642 -0.077342 -0.026328 -0.089630 -0.096198  0.026550   
27  -0.023677 -0.044642  0.059541 -0.040099 -0.042848 -0.043589  0.011824   
28   0.052606 -0.044642 -0.021295 -0.074528 -0.040096 -0.037639 -0.006584   
29   0.067136  0.050680 -0.006206  0.063187 -0.042848 -0.095885  0.052322   
..        ...       ...       ...       ...       ...       ...       ...   
412  0.074401 -0.044642  0.085408  0.063187  0.014942  0.013091  0.015505   
413 -0.052738 -0.044642 -0.000817 -0.026328  0.010815  0.007141  0.048640   
414  0.081666  0.050680  0.006728 -0.004523  0.109883  0.117056 -0.032356   
415 -0.005515 -0.044642  0.008883 -0.050428  0.025950  0.047224 -0.043401   
416 -0.027310 -0.044642  0.080019  0.098763 -0.002945  0.018101 -0.017629   
417 -0.052738 -0.044642  0.071397 -0.074528 -0.015328 -0.001314  0.004460   
418  0.009016 -0.044642 -0.024529 -0.026328  0.098876  0.094196  0.070730   
419 -0.020045 -0.044642 -0.054707 -0.053871 -0.066239 -0.057367  0.011824   
420  0.023546 -0.044642 -0.036385  0.000068  0.001183  0.034698 -0.043401   
421  0.038076  0.050680  0.016428  0.021872  0.039710  0.045032 -0.043401   
422 -0.078165  0.050680  0.077863  0.052858  0.078236  0.064447  0.026550   
423  0.009016  0.050680 -0.039618  0.028758  0.038334  0.073529 -0.072854   
424  0.001751  0.050680  0.011039 -0.019442 -0.016704 -0.003819 -0.047082   
425 -0.078165 -0.044642 -0.040696 -0.081414 -0.100638 -0.112795  0.022869   
426  0.030811  0.050680 -0.034229  0.043677  0.057597  0.068831 -0.032356   
427 -0.034575  0.050680  0.005650 -0.005671 -0.073119 -0.062691 -0.006584   
428  0.048974  0.050680  0.088642  0.087287  0.035582  0.021546 -0.024993   
429 -0.041840 -0.044642 -0.033151 -0.022885  0.046589  0.041587  0.056003   
430 -0.009147 -0.044642 -0.056863 -0.050428  0.021822  0.045345 -0.028674   
431  0.070769  0.050680 -0.030996  0.021872 -0.037344 -0.047034  0.033914   
432  0

In [67]:
y = diabetes_df.iloc[:,-1]

In [71]:
y

0      151.0
1       75.0
2      141.0
3      206.0
4      135.0
5       97.0
6      138.0
7       63.0
8      110.0
9      310.0
10     101.0
11      69.0
12     179.0
13     185.0
14     118.0
15     171.0
16     166.0
17     144.0
18      97.0
19     168.0
20      68.0
21      49.0
22      68.0
23     245.0
24     184.0
25     202.0
26     137.0
27      85.0
28     131.0
29     283.0
       ...  
412    261.0
413    113.0
414    131.0
415    174.0
416    257.0
417     55.0
418     84.0
419     42.0
420    146.0
421    212.0
422    233.0
423     91.0
424    111.0
425    152.0
426    120.0
427     67.0
428    310.0
429     94.0
430    183.0
431     66.0
432    173.0
433     72.0
434     49.0
435     64.0
436     48.0
437    178.0
438    104.0
439    132.0
440    220.0
441     57.0
Name: disease_prog, Length: 442, dtype: float64

In [68]:
# Your code here:
diabetes_model = LinearRegression()

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [73]:
# Your code her:
diabetes_model = LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [72]:
diabetes_data_train = X.iloc[:-20,:]

diabetes_target_train = y.iloc[:-20]

diabetes_data_test = X.iloc[-20:,:]

diabetes_target_test = y.iloc[-20:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [75]:
# Your code here:
model.fit(diabetes_data_train, diabetes_target_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [76]:
# Your code here:
diabetes_predict = model.predict(diabetes_data_test)

#### Print your `diabetes_target_test` and compare with the prediction. 

In [79]:
print(diabetes_target_test)

422    233.0
423     91.0
424    111.0
425    152.0
426    120.0
427     67.0
428    310.0
429     94.0
430    183.0
431     66.0
432    173.0
433     72.0
434     49.0
435     64.0
436     48.0
437    178.0
438    104.0
439    132.0
440    220.0
441     57.0
Name: disease_prog, dtype: float64


In [78]:
# Your code here:
r2_score(diabetes_target_test, diabetes_predict)

0.5850753022690574

It means that the explanatory variables of our model are able to explain 58% of the target variable.

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

# Your explanation here:

No, it is not. Our data target test is comprised by the last 20 observations of our sample, so they are real observed values for our target variable. Then based on the regression model we have built from the train data, we are able to make predictions for the target variable, but they don't have to be the same as the observed values.
Indeed, it is even better, otherwise it would mean we are maybe incurring in overfitting. 

Another way to see it, is that since the R squared is not 100% there are going to be differences (errors) between the observed target values (of the test data) and the predicted ones.

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../images/statsmodels.png)

In [80]:
# Your code here:
import statsmodels.api as sm

In [86]:
diabetes_data_train = sm.add_constant(diabetes_data_train, prepend=False)
mod = sm.OLS(diabetes_target_train, diabetes_data_train)
res = mod.fit()
print(res.summary())

                            OLS Regression Results                            
Dep. Variable:           disease_prog   R-squared:                       0.512
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     43.16
Date:                Fri, 09 Aug 2019   Prob (F-statistic):           4.64e-58
Time:                        18:26:29   Log-Likelihood:                -2281.1
No. Observations:                 422   AIC:                             4584.
Df Residuals:                     411   BIC:                             4629.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
age            0.3035     61.286      0.005      0.9

/home/salva/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

# Your answers here:

1. We have evidence to reject the H0, since the probability of the F-statistic is really small. IN other words, the probability of incurring in error type I upon rejecting the H0 is extremely small, so are confident to say that at least one the variables of our model is statistically significant in explaining variance of the target variable.

2. Yes. Variables age, s1, s2, s3, s4 and s6. They are not statistically significative (at a 5% significance level) in explaining variations of our target variable.

3. I would probably drop the non significant explanatory variables, and see how the adjustes R squared evolves.

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [91]:
# Your code here:
auto = pd.read_csv("../data/auto-mpg.csv")

Look at the first 5 rows using the `head()` function:

In [92]:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [93]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [94]:
# Your code here:
auto.model_year.min() #ELdest model

70

In [95]:
auto.model_year.max() # Newer model

82

Check the dataset for missing values and remove all rows containing at least one missing value.

In [96]:
# Your code here:
auto.isna().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [97]:
auto.dropna(inplace=True)

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [99]:
# Your code here:
auto.cylinders.value_counts().to_frame()

cylinders
4        199
8        103
6         83
3          4
5          3

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [100]:
# Your code here:
auto.drop("car_name", axis=1, inplace=True)

In [105]:
X = auto.iloc[:,1:]

In [107]:
X.head()

cylinders  displacement  horse_power  weight  acceleration  model_year
0          8         307.0        130.0    3504          12.0          70
1          8         350.0        165.0    3693          11.5          70
2          8         318.0        150.0    3436          11.0          70
3          8         304.0        150.0    3433          12.0          70
4          8         302.0        140.0    3449          10.5          70

In [108]:
y = auto.iloc[:,0:1]

In [111]:
y.head()

mpg
0  18.0
1  15.0
2  18.0
3  16.0
4  17.0

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [127]:
auto_model = LinearRegression()

In [128]:
# Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [129]:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../images/r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [130]:
# Your code here:
y_predict = model.predict(X_train)

In [131]:
r2_score(y_train, y_predict)

0.8107227953093896

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [132]:
# Your code here:
y_test_predict = model.predict(X_test)

In [133]:
r2_score(y_test, y_test_predict)

0.7942349075428589

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [134]:
auto_model_90 = LinearRegression()

In [135]:
# Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=12)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [136]:
# Your code here:
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

Compute the predicted values and r squared score for our new model and new sample data

In [138]:
y_predict = model.predict(X_train)
r2_score(y_train, y_predict)

0.8135037960103886

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [140]:
# Your code here:
y_test_predict = model.predict(X_test)
r2_score(y_test, y_test_predict)

0.7644393459851583

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [ ]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [ ]:
# Your code here:


Fit the model and print the ranking

In [ ]:
# Your code here:


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [ ]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [ ]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.